File: javascript/src/crypto/xchacha20-poly1305.js
/**
 * XChaCha20-Poly1305加密模块
 * 实现XChaCha20-Poly1305加密和解密
 *
 * 注意：@noble/ciphers返回Uint8Array，需要转换为Buffer
 */

const crypto = require('crypto');
const { xchacha20poly1305 } = require('@noble/ciphers/chacha');

const NONCE_LENGTH = 24;

/**
 * XChaCha20-Poly1305加密
 * @param {Buffer} key - 加密密钥（32字节）
 * @param {Buffer} plaintext - 明文
 * @param {Buffer} nonce - 初始化向量（24字节）
 * @returns {Object} { ciphertext, authTag }
 */
function _toUint8Array(buf) {
  // @noble/ciphers 要求纯 Uint8Array，不能是 Buffer（Buffer 的 .buffer 可能是共享的 ArrayBuffer）
  return new Uint8Array(buf.buffer, buf.byteOffset, buf.length);
}

function encryptXChaCha20Poly1305(key, plaintext, nonce) {
  const cipher = xchacha20poly1305(_toUint8Array(key), _toUint8Array(nonce));
  const ciphertextUint8 = cipher.encrypt(_toUint8Array(plaintext));

  const ciphertext = Buffer.from(ciphertextUint8);

  // XChaCha20-Poly1305: 最后16字节是认证标签
  const ciphertextOnly = ciphertext.slice(0, plaintext.length);
  const authTag = ciphertext.slice(plaintext.length);

  return { ciphertext: ciphertextOnly, authTag };
}

/**
 * XChaCha20-Poly1305解密
 * @param {Buffer} key - 解密密钥（32字节）
 * @param {Buffer} ciphertext - 密文
 * @param {Buffer} nonce - 初始化向量（24字节）
 * @param {Buffer} authTag - 认证标签（16字节）
 * @returns {Buffer} 明文
 */
function decryptXChaCha20Poly1305(key, ciphertext, nonce, authTag) {
  const ciphertextWithTag = Buffer.concat([ciphertext, authTag]);

  const cipher = xchacha20poly1305(_toUint8Array(key), _toUint8Array(nonce));
  const plaintextUint8 = cipher.decrypt(_toUint8Array(ciphertextWithTag));

  return Buffer.from(plaintextUint8);
}

/**
 * 生成随机nonce
 * @returns {Buffer} 24字节随机nonce
 */
function generateNonce() {
  return crypto.randomBytes(NONCE_LENGTH);
}

module.exports = {
  NONCE_LENGTH,
  encryptXChaCha20Poly1305,
  decryptXChaCha20Poly1305,
  generateNonce
};

File: python/src/crypto/xchacha20_poly1305.py
"""
XChaCha20-Poly1305加密模块
实现XChaCha20-Poly1305加密和解密

注意：由于Python cryptography库的限制，本实现使用ChaCha20-Poly1305（12字节nonce）。
如果未来cryptography库支持XChaCha20，可以升级到XChaCha20-Poly1305（24字节nonce）。

JavaScript版本使用@noble/ciphers的XChaCha20-Poly1305（24字节nonce）。
"""

from cryptography.hazmat.primitives.ciphers.aead import ChaCha20Poly1305
import os

NONCE_LENGTH = 12  # ChaCha20-Poly1305使用12字节nonce（Python cryptography库限制）


def encrypt_xchacha20_poly1305(key: bytes, plaintext: bytes, nonce: bytes) -> tuple[bytes, bytes]:
    """
    ChaCha20-Poly1305加密（Python回退实现）

    Args:
        key: 加密密钥（32字节）
        plaintext: 明文
        nonce: 初始化向量（12字节）

    Returns:
        Tuple[bytes, bytes]: (密文, 认证标签)
    """
    cipher = ChaCha20Poly1305(key)
    ciphertext = cipher.encrypt(nonce, plaintext, None)
    # ChaCha20-Poly1305: 最后16字节是认证标签
    ciphertext_only = ciphertext[: len(plaintext)]
    auth_tag = ciphertext[len(plaintext) :]
    return ciphertext_only, auth_tag


def decrypt_xchacha20_poly1305(
    key: bytes, ciphertext: bytes, nonce: bytes, auth_tag: bytes
) -> bytes:
    """
    ChaCha20-Poly1305解密（Python回退实现）

    Args:
        key: 解密密钥（32字节）
        ciphertext: 密文
        nonce: 初始化向量（12字节）
        auth_tag: 认证标签（16字节）

    Returns:
        bytes: 明文
    """
    cipher = ChaCha20Poly1305(key)
    # 重组密文和认证标签
    ciphertext_with_tag = ciphertext + auth_tag
    plaintext = cipher.decrypt(nonce, ciphertext_with_tag, None)
    return plaintext


def generate_nonce() -> bytes:
    """
    生成随机nonce

    Returns:
        bytes: 12字节随机nonce
    """
    return os.urandom(NONCE_LENGTH)
